### PR TITLE
TCR rewrite to simplify and split into extended contracts

### PR DESCRIPTION
**Pull request by mikec**
moves some of the TCR functionality into non TCR specific contracts that extend registry. Change `TokenCuratedRegistry.sol` implementation to simplify implementation. Part of this simplification was also to make some changes to `IChallenge.sol` and `IChallengeFactory.sol`

### CODE REVIEW

**Comment by ewilz** [comment]
Haven't looked at the code, but I'd appreciate an interactive rebase to rename the first commit to something like "Add IRegistry interface" since the current message is misleading!

**Comment by mikec** [comment]
@emaG3m I fixed that commit name

File: contracts/Registry/TokenCuratedRegistry.sol [hunk 1]
```diff
@@ -1,59 +1,97 @@
...
 
-    if (itemCurationData.challengeAddress == address(0)) {
-      return false;
-    } else if (itemCurationData.challengeResolved) {
-      return false;
-    } else {
-      return true;
-    }
+  function _challengePassed(bytes32 id) internal view returns (bool) {
+    return _challengeEnded(id) && challenges[id].passed();
```
Comment (emaG3m) [inline contracts/Registry/TokenCuratedRegistry.sol]: I think the outcomes of this function should be as follows:
1. `revert`  Challenge does not exist: none ever existed or all previous challenges are resolved on registry.
2. `revert`  Challenge exists. Challenge has not ended.
3. `true`     Challenge exists. Challenge has ended. Challenge has passed. Challenge has not been resolved on the registry.
4. `false`   Challenge exists. Challenge has ended. Challenge has not passed. Challenge has not been resolved on the registry.

`2.` would check out as long as Challenge contract reverts on `passed()` if it's not ended yet. Wondering  if `1.` checks out. This means either `challenges[id].ended()` or `challenges[id].passed()` needs to revert when called on a `0x0x` address. Do you know if this is the case?

File: contracts/Registry/TokenCuratedRegistry.sol [hunk 2]
```diff
@@ -1,59 +1,97 @@
...
+  }
 
-    if (itemCurationData.challengeAddress == address(0)) {
-      return false;
-    } else if (itemCurationData.challengeResolved) {
-      return false;
-    } else {
-      return true;
-    }
+  function _challengePassed(bytes32 id) internal view returns (bool) {
```
Comment (ewilz) [inline contracts/Registry/TokenCuratedRegistry.sol]: I think the outcomes of this function should be as follows:

1. `revert` Challenge does not exist: none ever existed or all previous challenges are resolved on registry.
2. `revert` Challenge exists. Challenge has not ended.
3. `true` Challenge exists. Challenge has ended. Challenge has passed. Challenge has not been resolved on the registry.
4. `false` Challenge exists. Challenge has ended. Challenge did not pass. Challenge has not been resolved on the registry.


`2.` would check out as long as Challenge contract reverts on `passed()` if it's not ended yet. Wondering if `1.` checks out. This means either `challenges[id].ended()` or `challenges[id].passed()` needs to revert when called on a `0x0` address. Do you know if this is the case?

File: contracts/Registry/TokenCuratedRegistry.sol [hunk 3]
```diff
@@ -1,59 +1,97 @@
...
+  function inApplicationPhase(bytes32 id) public view returns (bool) {
+    return exists(id) && !isLocked(id);
   }
 
-  function remove(bytes32 id) public itemNotLockedInChallenge(id) {
-    delete itemsCurationData[id];
-    super.remove(id);
+  // internals...
+
+  function _redeemReward(bytes32 id) internal returns (uint reward) {
```
Comment (ewilz) [inline contracts/Registry/TokenCuratedRegistry.sol]: One thing to keep in mind with rewards structure is how it could be modularized. This current implementation isn't compatible with our futarchy implementation, where you might reap rewards after another challenge has already started.

**Review by ewilz — CHANGES_REQUESTED** [review]
Waiting on answered question.

File: contracts/Registry/TokenCuratedRegistry.sol (hunk 2 above)
Comment (mikec) [reply contracts/Registry/TokenCuratedRegistry.sol]: @emaG3m it is the case that calling anything on `challenges[id]` would revert if it's a `0x0` address

**Comment by mikec** [comment]
@emaG3m great point about the requires, I think it makes logical sense to revert if something related to a challenge is called, but the challenge (or item id) does not exist. Also think it's especially valuable if these are `public view` functions. I changed them to `public view` because there's really no reason for them to be internal anymore - anyone (or any other contract) could call them to get the state of the challenge from this contract.

**Comment by mikec** [comment]
@emaG3m as for the rewards structure, yes this implementation is opinionated about the applicant/challenger stake matching. Also, the implementation can handle multiple challenges at a time, but only 1 challenge per item at a time. Not sure why this would not be compatible with the futarchy implementation?

**Comment by ewilz** [comment]
@mikec As I'm looking at the code it actually looks like it ISN'T opinionated about applicant/challenger stake matching (which IMO is a good thing since I think we want this to match goldin's vanilla version).

This may not be compatible with futarchy bc rewards still need to be dispersed long after calling `resolveChallenge`.  But if this could easily be done straight from the `Challenge` contract, maybe it transfers some rewards responsibility away from the `Registry`.

Either that, or there could be an extension that tracks `challenges` history in the regsitry contract, so you can interact with past challenges still. Goldin's implementation does this at the moment.

File: contracts/Challenge/IChallengeFactory.sol [hunk 4]
```diff
@@ -0,0 +1,5 @@
+pragma solidity ^0.4.24;
+
+interface IChallengeFactory {
+  function create(address registry, address challenger, address applicant) returns (address challenge);
```
Comment (ewilz) [inline contracts/Challenge/IChallengeFactory.sol]: why did you change the argument to `applicant`? `itemOwner` makes a lot more sense to me

File: contracts/Registry/TokenCuratedRegistry.sol [hunk 5]
```diff
@@ -1,59 +1,103 @@
...
+  // challenge fails.
+  function inApplicationPhase(bytes32 id) public view returns (bool) {
+    require(exists(id));
+    return !isLocked(id);
   }
 
-  function remove(bytes32 id) public itemNotLockedInChallenge(id) {
-    delete itemsCurationData[id];
-    super.remove(id);
+  // Returns true if the challenge for the given item id has passed.
```
Comment (ewilz) [inline contracts/Registry/TokenCuratedRegistry.sol]: wondering why you don't document `false` outcome

File: contracts/Registry/TokenCuratedRegistry.sol [hunk 6]
```diff
@@ -1,59 +1,103 @@
...
-    bytes32 id = keccak256(data);
-    itemsCurationData[id] = ItemCurationData(now + applicationPeriod, false, address(0), false);
-    return super.add(data);
+  // Returns true if the item exists and is not locked. We know that locked items are in
+  // the application phase, because the unlock time is set to now + applicationPeriod when
+  // items are added. Also, unlock time is set to 0 if an item is challenged and the
+  // challenge fails.
+  function inApplicationPhase(bytes32 id) public view returns (bool) {
+    require(exists(id));
+    return !isLocked(id);
```
Comment (ewilz) [inline contracts/Registry/TokenCuratedRegistry.sol]: I would think this line would need to be `return isLocked(id);` (which would have it's own bugs). But right now this would return `true` for a whitelisted item, no?

Comment (mikec) [reply contracts/Registry/TokenCuratedRegistry.sol]: oops, good catch!

Comment (ewilz) [reply contracts/Registry/TokenCuratedRegistry.sol]: ok even if you do correct this, this will return `true` for whitelisted items locked in a challenge

Comment (ewilz) [reply contracts/Registry/TokenCuratedRegistry.sol]: oh or maybe not.... since a challenge doesn't lock necessarily....which is...interesting

Comment (mikec) [reply contracts/Registry/TokenCuratedRegistry.sol]: @emaG3m this is fixed

File: contracts/Registry/TokenCuratedRegistry.sol (hunk 5 above)
Comment (mikec) [reply contracts/Registry/TokenCuratedRegistry.sol]: @emaG3m I added these to comments

File: contracts/Challenge/IChallengeFactory.sol (hunk 4 above)
Comment (mikec) [reply contracts/Challenge/IChallengeFactory.sol]: @emaG3m that makes sense, I changed it back

**Comment by mikec** [comment]
@emaG3m when you say `This may not be compatible with futarchy bc rewards still need to be dispersed long after calling resolveChallenge.`, you're talking about the funding that goes into the markets and basically doesn't come back to the funder until the markets are finally resolved, which happens after the decision period has already ended and resolveChallenge has been called? I think there are a few ways it could be made compatible with a futarchy challenge .. let me know what you think of either of these, and if I'm understanding correctly:

1. we could have futarchy challenge `reward()` return 0, and have the futarchy challenge contract handle all reward dispersement (I think this is what you were saying in your comment)
2. we could have futarchy challenge treat the transferred stake differently by holding it rather than using it to fund markets. market funding could happen in some other way (like via minting of FCR tokens, as martin suggested a couple weeks back)

I think if there's a case where `reward()` might return 0, we should be able to handle this .. right now it would revert. Should we allow this you think?

I'd like to avoid having to store state for past challenges if we can. Challenge contracts will still exist after their resolved and removed from registry state. Hopefully it will work for various challenge implementations to handle any additional reward logic after challenge resolution in the challenge contract itself.

**Comment by ewilz** [comment]
I think we can decide later on how it will work with futarchy. Tracking challenge history can always be an extension in the future if needed, no need to add that in now. Lmk give the code one last look!

File: contracts/Registry/StakedRegistry.sol [hunk 7]
```diff
@@ -1,40 +1,48 @@
...
-  }
+  mapping(bytes32 => uint) public ownerStakes;
 
   constructor(ERC20 _token, uint _minStake) public {
     token = _token;
     minStake = _minStake;
   }
 
-  function add(bytes32 data) public returns (bytes32) {
+  function add(bytes32 data) public returns (bytes32 id) {
```
Comment (ewilz) [inline contracts/Registry/StakedRegistry.sol:23]: One unfortunate drawback is that unlike goldin's implementation, you cannot add a custom amount. With goldin you `apply(item, stakeAmount, data)`. Another argument is that `data`, which might be important? (people implementing TCR with a UI might have an opinion on how important that data arg is).

If we wanted it to match one way around could be:

```
TokenCuratedRegistry {
  function add(bytes32 item, uint amount, string data) public returns(bytes32 id) {
     require(amount >= minStake);
      id = super.add(item);
      increaseStake(id, stakeAmount - minstake);
      // insert code here to store data
  }
}
```

Thinking out loud...maybe we can make these adjustments as needed later.

File: contracts/Registry/TokenCuratedRegistry.sol [hunk 8]
```diff
@@ -1,59 +1,107 @@
...
-    uint applicationExpiry;
-    bool whitelisted;
-    address challengeAddress;
-    bool challengeResolved;
+  // Creates a new challenge for an item.
+  // Requires that the item exists, and that there is no existing challenge for the item.
+  // Requires msg.sender (the challenger) to match the owner's stake by transferring to
+  // this contract. The challenger's and owner's stake is transferred to the newly created
+  // challenge contract.
+  function challenge(bytes32 id) public {
```
Comment (ewilz) [inline contracts/Registry/TokenCuratedRegistry.sol:44]: in the vanilla version, challenge should auto-pass if the `ownerStake` is less than `minStake` (`minStake` may have changed since item was added)

File: contracts/Registry/TokenCuratedRegistry.sol [hunk 9]
```diff
@@ -1,59 +1,107 @@
...
+  // is an ended challenge for the item.
+  function resolveChallenge(bytes32 id) public {
+    require(challengeEnded(id));
+    uint reward = challenges[id].reward();
+    require(token.transferFrom(challenges[id], this, reward));
+    if (challengePassed(id)) {
+      // if the challenge passed, reward the challenger (via token.transfer) and remove
+      // the item.
+      require(token.transfer(challenges[id].challenger(), reward));
+      delete ownerStakes[id];
```
Comment (ewilz) [inline contracts/Registry/TokenCuratedRegistry.sol]: `super.remove()` already does this.

File: contracts/Registry/TokenCuratedRegistry.sol [hunk 10]
```diff
@@ -1,59 +1,107 @@
...
+      super.remove(id);
```
Comment (ewilz) [inline contracts/Registry/TokenCuratedRegistry.sol:62]: should probably just be `remove(id)` without the super

File: contracts/Registry/LockableItemRegistry.sol [hunk 11]
```diff
@@ -0,0 +1,19 @@
+pragma solidity ^0.4.24;
+
+import './OwnedItemRegistry.sol';
+
+// provides a mapping of unlock times for items. Only allows
+// item removal when the unlock time has been exceeded.
+contract LockableItemRegistry is OwnedItemRegistry {
```
Comment (ewilz) [inline contracts/Registry/LockableItemRegistry.sol]: why does this need `OwnedItemRegistry` logic? not seeing it used anywhere.

Also I'm thinking something like `TimelockableItemRegistry` is a better name. Lockable sounds more general like you can lock/unlock items more at will.

File: contracts/Registry/TokenCuratedRegistry.sol [hunk 12]
```diff
@@ -1,59 +1,107 @@
...
+    } else {
+      // if the challenge failed, reward the applicant (by adding to their staked balance)
+      ownerStakes[id] = ownerStakes[id].add(reward).sub(minStake);
+    }
+    delete unlockTimes[id];
```
Comment (ewilz) [inline contracts/Registry/TokenCuratedRegistry.sol:67]: in the vanilla version, this would still be in application phase. I think keeping these lock times is useful for knowing what order things got added to the list. I don't see a point in deleting them anyway

File: contracts/Registry/TokenCuratedRegistry.sol [hunk 13]
```diff
@@ -1,59 +1,107 @@
...
+    require(token.transferFrom(challenges[id], this, reward));
+    if (challengePassed(id)) {
+      // if the challenge passed, reward the challenger (via token.transfer) and remove
+      // the item.
+      require(token.transfer(challenges[id].challenger(), reward));
+      delete ownerStakes[id];
+      super.remove(id);
+    } else {
+      // if the challenge failed, reward the applicant (by adding to their staked balance)
+      ownerStakes[id] = ownerStakes[id].add(reward).sub(minStake);
```
Comment (ewilz) [inline contracts/Registry/TokenCuratedRegistry.sol:65]: if minStake has changed since the challenge began, our accounting is going to be off here. We tracked this in the original version in the challengeStruct. I'm wondering how much some extra data storage would help with little things like this.

File: contracts/Registry/StakedRegistry.sol (hunk 7 above)
Comment (mikec) [reply contracts/Registry/StakedRegistry.sol:23]: For the `data` param, maybe we should consider adding this to an extension. Not sure the base class should be opinionated about including a data string.

For the `add()` function with additional stake, we could include a new function, so that `StakedRegistry` still implements the `IRegistry` function `add()`. Maybe something like this:

```
function addWithStake(bytes32 data, uint stakeAmount) public returns(bytes32 id) {
  require(stakeAmount > minStake);
  require(token.transferFrom(msg.sender, this, stakeAmount.sub(minStake)));
  id = add(data);
}
```

File: contracts/Registry/TokenCuratedRegistry.sol (hunk 8 above)
Comment (mikec) [reply contracts/Registry/TokenCuratedRegistry.sol:44]: @emaG3m we might want to hold off on implementing this until we add something like the "parameterizer", since until we do that it's not a possible condition

File: contracts/Registry/TokenCuratedRegistry.sol (hunk 10 above)
Comment (mikec) [reply contracts/Registry/TokenCuratedRegistry.sol:62]: I think we have to use super because of the require checks in the `remove()` function from this contract. It's supposed to be calling the `remove` from `StakedRegistry`

File: contracts/Registry/TokenCuratedRegistry.sol (hunk 12 above)
Comment (mikec) [reply contracts/Registry/TokenCuratedRegistry.sol:67]: @emaG3m if the item is in the application phase, and it's challenged and the challenge fails, wouldn't it make sense to take the item out of the application phase?

File: contracts/Registry/TokenCuratedRegistry.sol (hunk 13 above)
Comment (mikec) [reply contracts/Registry/TokenCuratedRegistry.sol:65]: @emaG3m that's true, but there's no way to change minStake in this implementation. I think we should deal with this issue when we add a way to change minStake

File: contracts/Registry/LockableItemRegistry.sol (hunk 11 above)
Comment (mikec) [reply contracts/Registry/LockableItemRegistry.sol]: @emaG3m you're right, it doesn't need to extend this. Fixed that. Also changed the name to `TimelockableItemRegistry`

**Review by ewilz — APPROVED** [review]
